clarify that contentSchema holds a subschema and when/how it applies

gregsdennis opened the pull request:
<!-- Love json-schema? Please consider supporting our collective:
👉  https://opencollective.com/json-schema/donate -->

<!-- In order to keep off topic discussion to a minimum, it helps if the "work to be done" is already agreed on. -->
<!-- Ideally, a GitHub Issue with the label `Status: Consensus` will have been concluded, and linked to in this PR. -->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

### What kind of change does this PR introduce?

<!-- E.g. a bugfix, feature, refactoring, etc… -->
clarification

### Issue & Discussion References

<!-- Pick at least one of the below options, and remove those which don't apply. -->
-  Closes #1381 <!-- Replace ___ with the issue number this PR resolves -->
-  Related to #1288 <!-- Use when the PR doesn't completely resolve an issue -->

### Summary

<!-- Explain the motivation for making this change. What existing problem does the pull request solve? -->

Updates the text for `contentSchema` to indicate that its value is indeed a subschema (and therefore should be treated as such when scanning for identifiers).  Also cleans up language around its dependency on `contentMediaType`.

I didn't include any explicit text about it containing identifiers.  Instead I declare that the value _is_ a subschema and removed the "SHOULD ignore" text discussed in the issue.

Also pertinent to the issue discussion is the final couple sentences (already present), which I ~broke out into a new paragraph~ rewrote to make it more apparent that it is a note of guidance rather than a requirement.

> Accessing the schema through the schema location IRI included as part of the
> annotation will ensure that it is correctly processed as a subschema. Using the
> extracted annotation value directly is only safe if the subschema is an embedded
> resource with both `$schema` and an absolute IRI `$id`.

### Does this PR introduce a breaking change?

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

No.

karenetheridge commented on specs/jsonschema-validation.md:
```diff
@@ -532,19 +532,24 @@ defined by [RFC 2046](#rfc2046).
+
+Accessing the schema through the schema location IRI included as part of the
+annotation will ensure that it is correctly processed as a subschema. Using the
+extracted annotation value directly is only safe if the subschema is an embedded
```
"safe" and "correctly processed as a subschema" is vague -- can we say something else? I think this is trying to say that the evaluation behaviour won't be reproducable because the schema is evaluated in isolation, rather than in the context of the surrounding dialect and location identifier (from the containing schema's $schema and $id keywords).  So how about instead saying something like:

> Because this subschema is intended to be processed in isolation, outside of the context of its containing schema, usage of both the `$schema` and `$id` keywords is recommended to ensure predictable and reproducable results.

gregsdennis replied:
This text was already present.  I just put it in a new paragraph.  I did think it was a bit convoluted.  Happy to update it.

gregsdennis replied:
I do think it could be okay to use the subschema _in_ its original context, so that should still be addressed.  This is what the original discussion was about.

jviotti commented on specs/jsonschema-validation.md:
```diff
@@ -532,19 +532,24 @@ defined by [RFC 2046](#rfc2046).
+
+Accessing the schema through the schema location IRI included as part of the
+annotation will ensure that it is correctly processed as a subschema. Using the
+extracted annotation value directly is only safe if the subschema is an embedded
```
Maybe we should make this a SHOULD?

gregsdennis replied:
I'm not sure what you're suggesting. This is informative, not a requirement.

jviotti replied:
So from what I'm reading here, there are edge cases in `contentSchema` if the schema doesn't have `$id` and `$schema`. If that's the case, shouldn't we highlight in more of a "SHOULD" manner, which from what I understand, it something you should do rather than a MUST (something you HAVE to do)? Or maybe "RECOMMENDED" is the right one similar to this case: https://json-schema.org/draft/2020-12/json-schema-validation#section-7.2.2-5?

gregsdennis replied:
It's not an edge case.  It just means that if you intend to use the `contentSchema` subschema solely in its own context, as you would if you received it as an annotation, then any relative `$ref`s will only be resolvable if the subschema has both `$schema` and `$id`.

This goes back to my comment [here](https://github.com/json-schema-org/json-schema-spec/issues/1381#issuecomment-2433652911) where I show a `contentSchema` subschema attempting to reference a definition in its parent schema.  If you extract the subschema (again, because you've received it as an annotation), then that reference fails.

There's not a best practice here.  Both approaches have valid use cases, and schema authors are free to do what makes sense for them.  This is merely a caution to schema authors to understand the implications of their approach.

gregsdennis replied:
It's obvious to me from @karenetheridge's and your comments that this paragraph isn't clear, so I'll reword it.

gregsdennis commented:
@karenetheridge / @jviotti I've rewritten the last paragraph note and added another "editor" footnote that points back to the subject issue linked above.  Let me know what you think.

gregsdennis commented on specs/jsonschema-validation.md:
```diff
@@ -532,19 +532,24 @@ defined by [RFC 2046](#rfc2046).
+The value of this property MUST be a valid JSON schema. The subschema is
+produced as an annotation.
+
+Since `contentMediaType` is required to provide instruction on how to interpret
+string content, the annotation schema produced by this keyword has no meaning if
+`contentMediaType` is not present.
```
Does that then also mean that identifiers are not to be processed in that case?

jdesrosiers replied:
I think it should be a "MUST".

gregsdennis commented:
@karenetheridge @jdesrosiers @jviotti I believe I've addressed your concerns here.

jviotti reviewed (approved):
Looks good to me

gregsdennis commented:
Closing and reopening to rerun build

jdesrosiers commented on specs/jsonschema-validation.md:
```diff
@@ -533,19 +533,32 @@ defined by [RFC 2046](#rfc2046).
+Note that evaluating the `contentSchema` subschema in-place (i.e. as part of its
+parent schema) will ensure that it is correctly processed. Independent use of
+the extracted subschema (as returned in an annotation) is only safe if the
+subschema is an embedded reource which defines both a `$schema` and an absolute
```
```suggestion
subschema is an embedded resource which defines both a `$schema` and an absolute
```

jdesrosiers commented on specs/jsonschema-validation.md:
```diff
@@ -533,19 +533,32 @@ defined by [RFC 2046](#rfc2046).
+string content, `contentSchema` SHOULD NOT produce an annotation if
+`contentMediaType` is not present.
+
+Note that evaluating the `contentSchema` subschema in-place (i.e. as part of its
+parent schema) will ensure that it is correctly processed. Independent use of
```
My initial reaction was that this doesn't make sense. The annotation is just the subschema. It's no longer in-place. It doesn't include the context of where it came from. So, how can it be evaluated in-place? Then it occurred to me that an annotation includes not just it's value, but also the schema location it came from and that location can be used to evaluate the `contentSchema` in-place. I don't think most readers are going to be knowledgeable enough to make that leap. This could use some clarification.

gregsdennis replied:
Does the footnote (`[^7]`) that follows not provide that clarity?

jdesrosiers replied:
No, that's not the thing I'm saying needs to be clarified. We say that the annotation is the subschema. We say that the subschema shouldn't be evaluated out of context from where it appeared in the schema and we explain why in footnote 7. What we don't explain is given a subschema without its parent context, how is it even possible to evaluate it in context. The value of the annotation is just the subschema, not the context. We can't evaluate the subschema in context because we don't know the context in which it needs to be evaluated. I hope that makes sense this time.

Of course the solution is that the location of the annotation keyword in the schema is how you know the context, but that's not intuitive. This is the only annotation where the location of the keyword in the schema is useful or necessary to know. Usually, we only care about the *value* of the annotation. In this case, we need to know the value and the schema location of the annotation. Actually, when used correctly (in context), the value of the annotation is useless and it's the schema location that the user actually uses.

gregsdennis replied:
> This is the only annotation where the location of the keyword in the schema is useful or necessary to know. Usually, we only care about the value of the annotation.

This is incorrect.  Annotation location has always been useful, especially in cases where you receive annotations from the same keyword in different locations, e.g. from `title`.  The location allows the consumer to decide which (or both/all) it wants to use.  This is Core, where annotations are defined.

> We can't evaluate the subschema in context because we don't know the context in which it needs to be evaluated.

How would we not know the context?  It's conveyed by the annotation location, which has always been defined to be a part of an annotation.

---

It's still not clear why you think that the existing text (including the lines following these) is insufficient.  It's saying, "Don't just evaluate this annotation value as a schema because it may rely on things that exist externally to it.  You probably need to evaluate it where it came from."

It's actually saying all of that, and then the footnote expands on that warning using an example.

jdesrosiers replied:
> > We can't evaluate the subschema in context because we don't know the context in which it needs to be evaluated.
>
> How would we not know the context? It's conveyed by the annotation location, which has always been defined to be a part of an annotation.

Yes, you're right. I acknowledged that in next paragraph. I was walking you through my thought process when I first read it and what I believe the vast majority of readers will be thinking when they read this section. If it took _me_ a minute to make that connection, most readers won't make it at all. Yes, the concept is unambiguously documented elsewhere, but most readers won't have every detail of JSON Schema memorized and I think this is a pretty esoteric detail.

jdesrosiers reviewed (changes requested):
Sorry, I seems this one slipped through the cracks for me a while ago.

jdesrosiers reviewed (approved):
I'm not going to fight with you about this. The text is correct even if I think readers will find it confusing. I'll approve the PR with or without the clarification I'm asking for.
